src/Renci.SshNet/SshCommand.cs

```diff
@@ -335,8 +358,62 @@ public string Execute()
             return EndExecute(BeginExecute(null, null));
         }
 
+#if FEATURE_TAP
+        /// <summary>
+        /// Asynchronously run a command.
+        /// </summary>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to observe.</param>
+        /// <param name="factory">The <see cref="System.Threading.Tasks.TaskFactory">TaskFactory</see> used to create the Task</param>
+        /// <param name="creationOptions">The TaskCreationOptions value that controls the behavior of the
+        /// created <see cref="T:System.Threading.Tasks.Task">Task</see>.</param>
+        /// <param name="scheduler">The <see cref="System.Threading.Tasks.TaskScheduler">TaskScheduler</see>
+        /// that is used to schedule the task that executes the end method.</param>
+        /// <returns>Returns <see cref="Task{Int32}"/> as command execution status result.</returns>
+        public Task<int> ExecuteAsync(
+            CancellationToken cancellationToken = default(CancellationToken),
+            TaskFactory<int> factory = null,
+            TaskCreationOptions creationOptions = default(TaskCreationOptions),
+            TaskScheduler scheduler = null)
+        {
+            return AsyncExtensions.FromAsync(
```

Comment:
This is how I propose to implement it:
```csharp
public async Task<int> ExecuteAsync(CancellationToken cancellationToken)
{
    var ctr = cancellationToken.Register((c) => ((IChannelSession)c).Dispose(), _channel, false);
    try
    {
        return await Task<int>.Factory.FromAsync(BeginExecute(), EndExecuteWithStatus).ConfigureAwait(false);
    }
    catch (Exception) when (cancellationToken.IsCancellationRequested) // ToDo: Define a more specific exception
    {
        throw new OperationCanceledException("Command execution has been cancelled.", cancellationToken);
    }
    finally
    {
        ctr.Dispose();
    }
}
```
No extensions required and it should handle cancellation correctly, Please note that this has not been tested.

Reply:
Ok yeah, the above code did not work. I have re-written it using a modified version of the extension which works correctly. To make it work you still need to set up a second task for the cancelation operation so you can determine if the task got cancelled, but this new version uses a callback to cancel the operation (closes the channel) and it works. Since we are actually closing the channel and terminating the original operation. I have verified the channel is being properly closed with this iteration.

I did leave the code as an extension as it would be useful to do use the same for other operations? But if you want I can simply remove the extension and implement the same code in the function itself? Either way the approach taken to set up a separate cancellation task so you know if the cancellation is needed is what works. Otherwise the original task just blocks until it's fully complete and the cancellation token does nothing.

Reply:
I think you are missing the point about cancellation: cancellation in dotnet is cooperative. If the underlying code doesn't observe it, the task will not get canceled. This is what I was proposing to cancel the channel in case of cancellation. There might be better options in SSH protocol, but this could do for now.

Reply:
Right, but cancelling the channel itself did not seem to early out of the await, which I think is the primary reason for the code I found and used. It creates a separate task that will bail when the cancellation token fires, and you can then dispose of the channel and return from the await. When I tried just closing the channel when the cancelation token fired, the await did not return until the /bin/sleep 5 completed. The way it is written now it works. The awaited task returns as soon as the cancelation token fires, and the channel also gets properly closed.

Reply:
Yes, but it is very "cludgy". If disposing the channel doesn't terminate the wait in EndExecute we have a problem that we need to solve first. I wouldn't want to layer patch-fixes on top to hide this problem.

Reply:
Ok I debugged it again. Part of the problem when I was debugging it last time was that the registration on the cancelation token was done before the command was started, so since I have such a short timeout in my test (100ms) by the time I debug that code it runs immediately, but there is no channel to close. So I changed it slightly to this:

```C#
            var asyncResult = BeginExecute();
            var ctr = cancellationToken.Register(_ => _channel?.Dispose(), false);
            try
            {
                return await Task<int>.Factory.FromAsync(asyncResult, EndExecuteWithStatus).ConfigureAwait(false);
            }
            catch (Exception) when (cancellationToken.IsCancellationRequested) // ToDo: Define a more specific exception
            {
                throw new OperationCanceledException("Command execution has been cancelled.", cancellationToken);
            }
            finally
            {
                ctr.Dispose();
            }
```

Now when it fires it has a channel to dispose and I can see that it is terminating the command and the await does actually finish and return as soon as the cancellation occurs, so that part was incorrect on my part. But the real problem is there is no cancellation exception happening so the await just returns and we never get into the catch block. In the test, it comes out and I get a blank string for the result, so it fails the test as I never got a cancellation exception.

With the way I have it structured in this PR, the channel is dispose and closed, the original await does finish but because we also run waited on another task to capture just the cancellation we now get the correct exception so the caller knows the command was cancelled.

I am not sure how else you would get the original await to throw the correct exception without doing that, since TaskFactory.FromAsync does not natively support a cancellation token.

Reply:
The other way this could potentially work is if there was a way to flag the running task that it was cancelled, and then you wake it up by setting the wait handle it's waiting it. It sees that it has been cancelled and then throws the appropriate exception that we want to indicate it was cancelled?